flocker/node/_protocol.py

```diff
@@ -0,0 +1,120 @@
+"""
+Communication protocol between control service and convergence agent.
+"""
+
+from twisted.protocols.amp import Argument, Command, Integer, String
+
+from .._model import NodeState, Deployment
+from .__config import current_from_configuration, marshal_configuration
+
+
+class NodeStateArgument(Argument):
+    """
+    AMP argument that takes a ``NodeState`` object.
+    """
+    def fromString(self, in_bytes):
+        #return json.loads(in_bytes) then decode NodeState
+        pass
+
+    def toString(self, node_state):
+        #return json.dumps(marshal_configuration(node_state))
+        pass
+
+
+# XXX This should suffice for both both communicating global current state
+# and the deployment configuration from control service to convergence
+# agent. We don't need to transfer used ports because the control service
```

Comment:
We might need to transfer used ports because ports might be used for some reason other than that they've been assigned to a Flocker-managed Docker container.  For example, port 22 is **probably** in use on any node you're going to deploy to so the control service should reject configuration changes that try to publish that port from a container.  I don't know of a good alternative to having the nodes report this information in order to handle this case.